Use identity for adding and updating hosts

thearifismail opened the pull request:
## Overview

This PR is being created to address [RHCLOUD-11482](https://issues.redhat.com/browse/RHCLOUD-11482).

It is about using `identity` when `adding` or `updating` a `host` in the `insights-inventory`.  When a Kafka message is received, its identity is retrieved from the message's platform_metadata.  If the identity type is `User`, then the host is saved and subsequent access to this host is granted using RBAC.  If the identity type is `System`, then the host's `system_profile['owner_id']` must match the "CN" in the `identity['system']['cn']`.  If the `owner_id` does not match the "CN", then an error is raised stating, "The owner in host does not match the owner in identity".

When adding a host, the `identity type` is `system` but `host['system_profile']['owner_id']` is not provided, then the `owner_id` is set to the `identity['system']['cn']`.  Once a system has been created, it is accessible only when the "CN" in the `identity` matches the `owner_id`. 

The hosts reported by `rhsm-conduit` do not provide any `identity`.  In that case, the `host['system_profile']['owner_id']` is set to the "subscription_manager_id".  If a message is received without an `identity` and the host data is missing `subscription_manager_id`, the host is rejected.  If a message is received with no identity and the reporter is not rhsm-conduit, then the host is rejected.

This PR has a lot of code changes.  It is suggested to start the review from the `handle_message` method in `app/queue/queue.py` and follow the `add_host` and `find_existing_host` methods.  Among tests, major change is the use "User" and "System" as `identity types` and the use the use of`b64_identity` in `platform_metadata`.

This PR is continuation of the [closed PR #802](https://github.com/RedHatInsights/insights-host-inventory/pull/802), which had way too many commits.  To avoid complexities arising from merging multiple commits, this PR has been created.

I hope this helps.

## Secure Coding Practices Checklist GitHub Link

- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist

- [x] Input Validation
- [ ] Output Encoding
- [x] Authentication and Password Management
- [ ] Session Management
- [x] Access Control
- [ ] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [x] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

kruai commented on app/utils.py:
```diff
@@ -188,6 +188,15 @@ def to_json(self):
+        return f"HostWrapper({json.dumps(self.__data, indent=2)})"
+
+    def __eq__(self, other):
+        if isinstance(other, dict):
+            return self.__data == other
+        else:
+            return self.__data == other.__data
```
It seems strange to implement a comparator that compares against just any dict. The `else` part of it I'm OK with, though again, I think it would break if you passed in any object that wasn't a dict or a HostWrapper

thearifismail replied:
If the object is not a HostWrapper instance, then it should not be called.  That said, the check `if isinstance(other, dict):` his being used to make a couple of tests pass.  IMO, these tests should scrapped because the use of identity has rendered them un-passable without this hack.  

For now, I am removing this hack and commenting out these bad tests.

kruai commented on tests/helpers/db_utils.py:
```diff
@@ -19,8 +20,12 @@
 DB_NEW_FACTS = {"newfact1": "newvalue1", "newfact2": "newvalue2"}
 
 
+log = logging.getLogger(__name__)
```
can we call this var `logger` for consistency with all the other instances?

thearifismail replied:
Done!

kruai commented on tests/helpers/test_utils.py:
```diff
@@ -144,3 +163,15 @@ def valid_system_profile():
+    return {
+        "request_id": "b9757340-f839-4541-9af6-f7535edf08db",
+        "archive_url": "http://s3.aws.com/redhat/insights/1234567",
+        "b64_identity": SYSTEM_API_KEY.decode("ascii"),
```
I'm a little confused... in the `SYSTEM_API_KEY` definition, it's encoded with UTF-8. Here's it's decoding from ASCII. I see UTF-8 and ASCII encoding/decoding in various places. Which one is it?

thearifismail replied:
reporters provide `utf-8` encoded `b64_identity`, which has to be decoded to ASCII in platform-metadata.  Without ASCII decoding, the apiKey remains in `bytes` format and is not `JSON serializable`.   To make it JSON serializable, it must be decoded to `ASCII`, otherwise `return json.dumps(message).encode("utf-8")` throws `TypeError: Object of type 'bytes' is not JSON serializable` error

kruai commented on tests/helpers/test_utils.py:
```diff
@@ -144,3 +163,15 @@ def valid_system_profile():
+    }
+
+
+def get_encoded_idstr():
+    return SYSTEM_API_KEY.decode("ascii")
```
If you define this function above `get_platform_metadata_with_system_identity()`, you can use its value for "b64_identity" (small refactoring)

thearifismail replied:
Done!

thearifismail replied:
Done!

kruai commented on tests/test_api_facts.py:
```diff
@@ -169,10 +169,14 @@ def test_put_facts_with_RBAC_denied(subtests, mocker, api_put, db_create_host, d
+    response_status, response_data = api_put(url, DB_NEW_FACTS, SYSTEM_IDENTITY)
 
-    assert_response_status(response_status, 200)
+    # with the use of identity, response_status 400 is the expected one.
+    assert_response_status(response_status, 400)
```
Are you sure that's what we should expect here? `HTTP 400` is for bad/malformed requests. If we're trying to say "permission denied", isn't `HTTP 401` more appropriate?

thearifismail replied:
Fixed the test to pass to use the correct identity

thearifismail replied:
Fixed the test.

kruai commented on tests/test_deduplication.py:
```diff
@@ -4,11 +4,14 @@
 from tests.helpers.db_utils import minimal_db_host
 from tests.helpers.test_utils import generate_uuid
 
+# from tests.helpers.test_utils import USER_IDENTITY
+
```
delete this

thearifismail replied:
Done!!!

thearifismail replied:
Gone!

kruai commented on tests/test_deduplication.py:
```diff
@@ -4,11 +4,14 @@
     fqdn = "fred.flintstone.com"
     canonical_facts = {"fqdn": fqdn, "bios_uuid": generate_uuid(), "rhel_machine_id": generate_uuid()}
 
+    # host = minimal_db_host(identity=USER_IDENTITY["identity"], canonical_facts=canonical_facts)
```
delete this

thearifismail replied:
Gone!!!

thearifismail replied:
Gone!

kruai commented on utils/kafka_producer.py:
```diff
@@ -5,6 +5,9 @@
 from kafka import KafkaProducer
 from ttictoc import TicToc
 
+# import rhsm_payloads as payloads
+# import user_payloads as payloads
+
```
delete this

thearifismail replied:
Gone!!!

thearifismail replied:
Gone!

kruai reviewed (changes requested):
I'm not sure I got to everything I wanted to comment on, but this is a huge PR and we might need to talk through it a bit. One of the biggest changes that caused ripples throughout the files is that the "identity" dict structure now has a top-level field called "identity". This required changes to be made across every file that used this structure, but I'm not yet convinced it was a necessary change.

I'm also nervous about using mock Identity values as the base for Identities we're actually using for hosts; these used to be in the test utils, but now they're being used for actual requests. It feels dangerous to use those as a base, especially when the values in the mocks look very believable, and would be way harder to track down than an account number named "test" or something like that.

Finally, I'm a little confused about our usage of the `Identity` class. It was defined and we use it in a few places, but most of the time, the code is just using the raw `dict` structure, and assuming all the keys are in the right place. I think if you leaned a little more on the class implementation, it could simplify the code in a lot of places.

thearifismail commented on app/auth/identity.py:
```diff
@@ -76,7 +76,13 @@ def _asdict(self):
 
     def __eq__(self, other):
-        return self.account_number == other.account_number
+        if isinstance(other, dict):
```
It is back to `    def __eq__(self, other):
        return self.account_number == other.account_number
`

thearifismail commented on app/auth/identity.py:
```diff
@@ -76,7 +76,13 @@ def _asdict(self):
+            else:
+                return self.account_number == other.get("account_number")
+        else:
+            return self.account_number == other.account_number
```
this snippet has been removed.

thearifismail commented on app/queue/queue.py:
```diff
@@ -46,6 +43,48 @@ class OperationSchema(Schema):
+    base64_bytes = base64_id.encode("ascii")
+    id_bytes = base64.b64decode(base64_bytes)
+    id = id_bytes.decode("ascii")
+    return json.loads(id)
```
Added exception handler

thearifismail replied:
Added exception handler

thearifismail commented on app/queue/queue.py:
```diff
@@ -46,6 +43,48 @@ class OperationSchema(Schema):
+    identity = None
+    # check the reporter
+    if not metadata.get("b64_identity") and not host.get("reporter") == "rhsm-conduit":
+        raise ValueError("Missing identity and the rhsm-conduit reporter, which does not use identity")
```
Fixed the comment and the message:
`def _get_identity(host, metadata):
    identity = None
    # Identity is required unless the reporter type is "rhsm-conduit"
    if not metadata.get("b64_identity") and not host.get("reporter") == "rhsm-conduit":
        raise ValueError("Provide identity or the reporter MUST be rhsm-conduit, which does not provide identity")
`

thearifismail replied:
Updated the comment and  message

thearifismail commented on app/queue/queue.py:
```diff
@@ -46,6 +43,48 @@ class OperationSchema(Schema):
 
+# input is a base64 encoded string and returns the identity dictionary
+def _decode_id(encoded_id):
+    base64_id = encoded_id
```
Fixed the code.

thearifismail commented on app/queue/queue.py:
```diff
@@ -46,6 +43,48 @@ class OperationSchema(Schema):
+        raise ValueError("Missing identity and the rhsm-conduit reporter, which does not use identity")
+
+    # rhsm report does not provide identity.  Set identity type to system for subsequent access
+    if not metadata.get("b64_identity") and host.get("reporter") == "rhsm-conduit":
```
This is not correct.  A better comment and message hopefully helps.

kruai replied:
The conditionals can still be refactored to the above logic to make the flow easier to follow (see the code block in my comment).

thearifismail commented on app/queue/queue.py:
```diff
@@ -136,11 +175,15 @@ def handle_message(message, event_producer):
 
-    # create a dummy identity for working around the identity requirement for CRUD operations
-    identity = Identity(USER_IDENTITY)
+    identity = _get_identity(validated_operation_msg.get("data"), platform_metadata)
+
+    host = validated_operation_msg["data"]
```
Fixed.

thearifismail commented on lib/identity.py:
```diff
@@ -0,0 +1,45 @@
+__all__ = ("get_system_cert_auth_identity", "get_system_classic_identity", "get_user_basic_auth_identity")
+
+
+SYSTEM_IDENTITY = {
+    "identity": {
+        "account_number": "sysaccount",
+        "auth_type": "cert-auth",
+        "internal": {"auth_time": 6300, "org_id": "3340851"},
+        "system": {"cert_type": "system", "cn": "1b36b20f-7fa0-4454-a6d2-008294e06378"},
+        "type": "System",
+    }
+}
+
+INSIGHTS_CLASSIC_IDENTITY = {
+    "identity": {
+        "account_number": "classic",
+        "auth_type": "classic-proxy",
+        "internal": {"auth_time": 6300, "org_id": "3340851"},
+        "system": {},
+        "type": "System",
+    }
+}
+
+USER_IDENTITY = {
+    "identity": {
+        "account_number": "usraccount",
+        "type": "User",
+        "auth_type": "basic-auth",
+        "user": {"email": "tuser@redhat.com", "first_name": "test"},
+    }
+}
```
The reason is that `b64_identity` passed in by the reporters has {"identity": {value}} and needs handling.

kruai replied:
That's true, but you coded the functions that decode it and return the values; why not just have them return the "identity" node so you don't have to fetch "identity" each time? Better yet, why not initialize an Identity object from it, since the class is already defined? I think that would be a lot cleaner than passing around a dictionary

thearifismail commented on lib/identity.py:
```diff
@@ -0,0 +1,45 @@
+        "account_number": "sysaccount",
+        "auth_type": "cert-auth",
+        "internal": {"auth_time": 6300, "org_id": "3340851"},
+        "system": {"cert_type": "system", "cn": "1b36b20f-7fa0-4454-a6d2-008294e06378"},
```
This is CN is primarily used for testing.  Reporters are supposed to provide their identities, except `rhsm-conduit`, which does not have any identity when reporting a host BUT for subsequent access, `system` identity will be used.  That's the only reason that it is being used in the `queue` module; even then the dummy `CN` is replaced using `system_profile['owner_id] = SUBSCRIPTION_MANAGER_ID`

kruai replied:
So... I understand this reasoning, but there are 2 sticking points for me:
- If these are values that should be used for testing, they should be defined in the test utils instead.
- If these variables are used in production, I'm nervous having the values pre-populated in case there's a logic error somewhere and we end up with mock data in our prod database

thearifismail commented on app/queue/queue.py:
```diff
@@ -136,11 +175,15 @@ def handle_message(message, event_producer):
 
-    # set account_number in dummy idenity to the actual account_number received in the payload
-    identity.account_number = validated_operation_msg["data"]["account"]
+    identity = Identity(identity.get("identity"))
```
It does seem so.  `_get_identity()` returns a dictionary with only one key `identity` at the top level.  That dictionary from _get_identity is used to create the identity object of type `Identity`

kruai replied:
It does return a dictionary, but... why doesn't it return an Identity? That would make a lot of this cleaner

thearifismail commented on app/queue/queue.py:
```diff
@@ -46,6 +43,48 @@ class OperationSchema(Schema):
+    return identity
+
+
+# test for no system_profile or no owner_id,
```
Fixed the comment.

lphiri commented:
Reviewed  with @thearifismail . The general approach looks good to me ( review did not cover test coverage).

kruai commented:
I've left some responses, though I really think @ryandillinfelton and/or @tahmidefaz should take a look as well... I want to make sure my comments aren't off-base!

thearifismail commented:
> I've left some responses, though I really think @ryandillinfelton and/or @tahmidefaz should take a look as well... I want to make sure my comments aren't off-base!

@kruai , All issue have been addressed.  The issues were:

1. refactored the checks for `identity` and `rhsm-conduit`
2. removed the `identity` keyword from the `identity` object.
3. removed the identity file from lib directory and put it back in `test_utils`.  For rhsm-reported systems, identity is created using a bare-bone dictionary with no `cn`.  The `cn` is set to the `subscription_manager_id` of the host.

kruai reviewed (changes requested):
In addition to my inline comments, I just have a general question about these sweeping changes to our MQ tests. Are 100% of messages read from the Kafka queue going to both be of type "System" and have an identity (besides RHSM)? Because if they don't, it's gonna start breaking for a lot of people

tahmidefaz commented on app/queue/queue.py:
```diff
@@ -136,11 +187,14 @@ def handle_message(message, event_producer):
+    host = validated_operation_msg["data"]
+    identity = _get_identity(host, platform_metadata)
+
+    # basic-auth does not need owner_id
+    if identity.get("type") == "System":
+        host = _set_owner(host, identity)
 
-    # set account_number in dummy idenity to the actual account_number received in the payload
-    identity.account_number = validated_operation_msg["data"]["account"]
+    identity = Identity(identity)
```
I agree with @kruai here, doing this will clear this PR up a little bit

thearifismail replied:
Addressed all issues mentioned in this group of comments, except `Enum`.  Will use a separate commit for `Enum` because it affects multiple files.

thearifismail replied:
IMO, using `Enum` for `IdentityType` and `AuthTypes` should wait for now and be added in a separate PR.

kruai replied:
It looks like you added AuthTypes, but I've created a follow-up Jira for IdentityTypes here https://issues.redhat.com/browse/RHCLOUD-12861

thearifismail commented on app/queue/queue.py:
```diff
@@ -32,11 +33,11 @@
-    "type": "User",
-    "user": {"email": "tuser@redhat.com", "first_name": "test"},
+SYSTEM_IDENTITY = {
+    "account_number": "sysaccount",
```
Gone!

tahmidefaz commented:
I do not have any additional comments at this point from what Asa mentioned. Good progress! We're almost there :+1:

thearifismail commented on app/queue/queue.py:
```diff
@@ -46,6 +47,56 @@ class OperationSchema(Schema):
+    try:
+        id = base64.b64decode(encoded_id)
+        decoded_id = json.loads(id)
+    except json.JSONDecodeError as jde:
+        raise json.JSONDecodeError(jde)
```
If a bad json is passed in, JSONDecodeError is raised instead of failure encountered somewhere down the line.

kruai replied:
What do you mean? You're catching a JSONDecodeError and re-raising that same error. This is effectively the same as not having the try/except clause at all

thearifismail commented on tests/test_api_facts.py:
```diff
@@ -169,8 +169,10 @@ def test_put_facts_with_RBAC_denied(subtests, mocker, api_put, db_create_host, d
 def test_put_facts_with_RBAC_bypassed_as_system(api_put, db_create_host, enable_rbac):
     host = db_create_host(
-        extra_data={"facts": DB_FACTS, "system_profile_facts": {"owner_id": SYSTEM_IDENTITY["system"]["cn"]}}
+        extra_data={"facts": DB_FACTS, "system_profile_facts": {"owner_id": SYSTEM_IDENTITY["system"].get("cn")}}
```
As the name indicates, the intent is to use "System" as account type to bypass RBAC.   Using `identity_type: system` requires `system_profile['ownd_id']` set to `cn`

thearifismail commented on tests/test_api_facts.py:
```diff
@@ -169,8 +169,10 @@ def test_put_facts_with_RBAC_denied(subtests, mocker, api_put, db_create_host, d
+        extra_data={"facts": DB_FACTS, "system_profile_facts": {"owner_id": SYSTEM_IDENTITY["system"].get("cn")}}
     )
+
+    host.account = SYSTEM_IDENTITY["account_number"]
```
A host created by the `db_create_host` fixtures has `account = `usraccount`, whereas `SYSTEM_IDENTITY["account_number"]` has  value `sysaccount`. That's why host.account is set to `sysaccount`

thearifismail commented on tests/test_api_hosts_delete.py:
```diff
@@ -216,6 +216,7 @@ def test_delete_host_with_RBAC_bypassed_as_system(
     api_delete_host, event_datetime_mock, event_producer_mock, db_get_host, db_create_host, enable_rbac
 ):
     host = db_create_host(extra_data={"system_profile_facts": {"owner_id": SYSTEM_IDENTITY["system"]["cn"]}})
+    host.account = SYSTEM_IDENTITY["account_number"]
```
Same reason as stated above.

thearifismail commented on tests/test_api_hosts_update.py:
```diff
@@ -444,6 +444,7 @@ def test_patch_host_with_RBAC_denied(
 
 def test_patch_host_with_RBAC_bypassed_as_system(api_patch, db_create_host, event_producer_mock, enable_rbac):
     host = db_create_host(extra_data={"system_profile_facts": {"owner_id": SYSTEM_IDENTITY["system"]["cn"]}})
+    host.account = "sysaccount"
```
same reason as above.

thearifismail commented on utils/payloads.py:
```diff
@@ -1,10 +1,22 @@
 from datetime import timedelta
 from datetime import timezone
 
+IDENTITY = {
+    "identity": {
+        "account_number": "sysaccount",
+        "type": "System",
+        "auth_type": "cert-auth",
+        "system": {"cn": "1b36b20f-7fa0-4454-a6d2-008294e06378", "cert_type": "system"},
+        "internal": {"org_id": "3340851", "auth_time": 6300},
+    }
+}
+apiKey = base64.b64encode(json.dumps(IDENTITY).encode("utf-8"))
+
```
This `identity` key is to simulate what `puptoo` sends as `b64_identity`

thearifismail replied:
Done!

thearifismail commented on utils/payloads.py:
```diff
@@ -511,11 +525,11 @@ def random_uuid():
 
 def build_host_chunk():
-    account = os.environ.get("INVENTORY_HOST_ACCOUNT", "test")
+    account = os.environ.get("INVENTORY_HOST_ACCOUNT", IDENTITY["identity"]["account_number"])
```
same reason as above.

kruai replied:
I still had the same question as above :)

thearifismail commented on utils/payloads.py:
```diff
@@ -532,9 +546,10 @@ def build_host_chunk():
+        # "subscription_manager_id": "1b36b20f-7fa0-4454-a6d2-008294e06378",
         "system_profile": create_system_profile(),
         "stale_timestamp": (datetime.now(timezone.utc) + timedelta(days=1)).isoformat(),
+        # "reporter": "rhsm-conduit",
```
To test a scenario using no-rhsm reporter.

thearifismail replied:
Changed to payloads.py are made based on what's being tests or was tested the last.   Need to update the test utility to not have to modify payloads.py

thearifismail commented on app/queue/queue.py:
```diff
@@ -32,11 +33,11 @@
-    "auth_type": "basic-auth",
-    "type": "User",
-    "user": {"email": "tuser@redhat.com", "first_name": "test"},
+SYSTEM_IDENTITY = {
+    "account_number": "sysaccount",
+    "auth_type": "cert-auth",
+    "system": {"cert_type": "system"},
+    "type": "System",
```
Now `_get_identity()` returns `Identity` instead of a dictionary

thearifismail commented on app/queue/queue.py:
```diff
@@ -46,6 +47,56 @@ class OperationSchema(Schema):
+
+
+def _get_identity(host, metadata):
+    identity = None
```
Gone!

thearifismail commented on app/queue/queue.py:
```diff
@@ -46,6 +47,56 @@ class OperationSchema(Schema):
+    # rhsm reporter does not provide identity.  Set identity type to system for access the host in future.
+    if not metadata.get("b64_identity"):
+        if host.get("reporter") == "rhsm-conduit":
+            SYSTEM_IDENTITY["account_number"] = host.get("account")
+            SYSTEM_IDENTITY["system"]["cn"] = host.get("subscription_manager_id")
+            identity = SYSTEM_IDENTITY
```
Done!

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -1008,35 +1032,12 @@ def test_other_values_are_ignored(value):
     assert True
 
 
-def test_handle_message_with_different_account(mocker, flask_app, subtests):
-    mocker.patch("app.queue.queue.build_event")
-    add_host = mocker.patch("app.queue.queue.add_host", return_value=(mocker.MagicMock(), None, None, None))
-
-    operation_raw = "🧜🏿‍♂️"
-    messages = (
-        f'{{"operation": "", "data": {{"display_name": "{operation_raw}{operation_raw}", "account": "dummy"}}}}',
-    )
-
-    identity = Identity(SYSTEM_IDENTITY)
-    identity.account_number = "dummy"
-
-    for message in messages:
-        with subtests.test(message=message):
-            host_id = generate_uuid()
-            add_host.reset_mock()
-            add_host.return_value = ({"id": host_id}, host_id, None, AddHostResult.updated)
-            handle_message(message, mocker.Mock())
-            add_host.assert_called_once_with(
-                {"display_name": f"{operation_raw}{operation_raw}", "account": "dummy"}, identity
-            )
-
-
```
This test was not valid anymore due to the way identity is used when adding a host.

kruai replied:
Ah... it was replaced with `test_add_host_with_wrong_owner`, right?

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -1058,10 +1059,119 @@ def test_handle_message_side_effect(mocker, flask_app):
+    )
+    message = wrap_message(host.data(), "add_host")
+
+    try:
+        handle_message(json.dumps(message), mock_event_producer)
+    except ValueError as ve:
+        print(str(ve))
```
Replaced `try/except` with `pytest.raises`

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -420,20 +440,19 @@ def test_add_host_with_owner_id(event_datetime_mock, mq_create_or_update_host, d
     """
     Tests that owner_id in the system profile is ingested properly
     """
-    owner_id = generate_uuid()
-    host = minimal_host(system_profile={"owner_id": owner_id})
+    host = minimal_host_owned_by_system(system_profile={"owner_id": OWNER_ID})
     created_host_from_event = mq_create_or_update_host(host)
     created_host_from_db = db_get_host(created_host_from_event.id)
-    assert created_host_from_db.system_profile_facts == {"owner_id": owner_id}
+    assert created_host_from_db.system_profile_facts == {"owner_id": OWNER_ID}
 
 
 def test_add_host_with_owner_incorrect_format(event_datetime_mock, mq_create_or_update_host, db_get_host):
     """
     Tests that owner_id in the system profile is rejected if it's in the wrong format
     """
     owner_id = "Mike Wazowski"
-    host = minimal_host(system_profile={"owner_id": owner_id})
-    with pytest.raises(ValidationException):
+    host = minimal_host_owned_by_system(system_profile={"owner_id": owner_id})
+    with pytest.raises(InventoryException):
```
Fixed!.

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -402,16 +410,28 @@ def test_add_host_key_filtering_system_profile(mq_create_or_update_host, db_get_
     }
 
 
+# def test_add_host_not_marshmallow_system_profile(mocker, mq_create_or_update_host):
+#     mock_deserialize_host = partial(mocker.Mock, wraps=deserialize_host)
+#     mock = mocker.patch("app.serialization.deserialize_host", new_callable=mock_deserialize_host)
+
+#     host_to_create = minimal_host_owned_by_system(system_profile={"number_of_cpus": 1})
+#     mq_create_or_update_host(host_to_create)
+#     mock.assert_called_once_with(mocker.ANY, MqHostSchema, None)
+
+#     assert type(MqHostSchema._declared_fields["system_profile"]) is not marshmallow.fields.Nested
+
+
```
Gone!

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -292,6 +275,30 @@ def test_add_host_with_system_profile(event_datetime_mock, mq_create_or_update_h
+
+    with pytest.raises(InventoryException):
+        key, event, headers = mq_create_or_update_host(host, return_all_data=True)
+        assert_mq_host_data(key, event, expected_results, host_keys_to_check)
```
Gone!

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -292,6 +275,30 @@ def test_add_host_with_system_profile(event_datetime_mock, mq_create_or_update_h
+
+    host = minimal_host_owned_by_system(insights_id=expected_insights_id, system_profile=expected_system_profile)
+
+    expected_results = {
+        "host": {**host.data()},
+        "platform_metadata": {},
+        "timestamp": timestamp_iso,
+        "type": "created",
+    }
```
Yep.  Gone! and removed other instances also. Now have `expected_results = {
        "host": {**host.data()},
    }
`

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -1058,10 +1059,119 @@ def test_handle_message_side_effect(mocker, flask_app):
+    message = wrap_message(host.data(), "add_host")
+
+    try:
+        handle_message(json.dumps(message), mock_event_producer)
```
Fixed.  It was a left over snippet which should have been replaced by `pytest.raises()`

kruai commented on app/auth/identity.py:
```diff
@@ -94,8 +100,22 @@ def validate(identity):
+            if identity.auth_type not in auth_types:
+                raise ValueError("The auth_type is invalid.")
+
+        elif identity.identity_type == "User":
```
The "System" and "User" identity types should also be enums, but this can wait for a follow-up PR.

kruai commented on utils/payloads.py:
```diff
@@ -1,10 +1,22 @@
+        "account_number": "sysaccount",
+        "type": "System",
+        "auth_type": "cert-auth",
+        "system": {"cn": "1b36b20f-7fa0-4454-a6d2-008294e06378", "cert_type": "system"},
```
Should use the enum value for cert_type

thearifismail replied:
`enum` values to be implemented in a separate PR.

kruai reviewed (changes requested):
Things are looking better! 
Since we're doing the `identity_type` enum as a follow-up, I've created a task for it here: https://issues.redhat.com/browse/RHCLOUD-12861

kruai commented on app/auth/identity.py:
```diff
@@ -94,8 +100,22 @@ def validate(identity):
+            if identity.auth_type not in auth_types:
+                raise ValueError("The auth_type is invalid.")
+
+        elif identity.identity_type == "User":
```
(update: created here https://issues.redhat.com/browse/RHCLOUD-12861)

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -177,36 +183,12 @@ def test_handle_message_failure_invalid_surrogates(mocker, display_name):
     add_host.assert_not_called()
 
 
-def test_handle_message_unicode_not_damaged(mocker, flask_app, subtests):
-    mocker.patch("app.queue.queue.build_event")
-    add_host = mocker.patch("app.queue.queue.add_host", return_value=(mocker.MagicMock(), None, None, None))
-
-    operation_raw = "🧜🏿‍♂️"
-    operation_escaped = json.dumps(operation_raw)[1:-1]
-
-    messages = (
-        f'{{"operation": "", "data": {{"display_name": "{operation_raw}{operation_raw}", "account": "test"}}}}',
-        f'{{"operation": "", "data": {{"display_name": "{operation_escaped}{operation_escaped}","account": "test"}}}}',
-        f'{{"operation": "", "data": {{"display_name": "{operation_raw}{operation_escaped}", "account": "test"}}}}',
-    )
-
-    for message in messages:
-        with subtests.test(message=message):
-            host_id = generate_uuid()
-            add_host.reset_mock()
-            add_host.return_value = ({"id": host_id}, host_id, None, AddHostResult.updated)
-            handle_message(message, mocker.Mock())
-            add_host.assert_called_once_with(
-                {"display_name": f"{operation_raw}{operation_raw}", "account": "test"}, Identity(USER_IDENTITY)
-            )
-
-
```
The messages structure was not the best, particularly wrapping the whole message in single quote.  Adding identity would increase the length beyond 120 chars, linters' limit for a line.  Rounding by linter produced bad message.  Until, I forced it work somehow.  Dues to the changed in identity code, the test did not work at all, so finally it was its to time to go.

kruai replied:
Whoa, this is confusing - did you edit my comment and replace its content with your response? :thinking: 
Regardless, a linting error in implementation does not constitute a test being invalid. If you're having trouble getting it to format correctly, please reach out to me (or your other teammates!) to get help

thearifismail commented on tests/test_api_auth.py:
```diff
@@ -100,7 +100,6 @@ def test_invalid_system_identities(flask_client, subtests):
             assert 401 == response.status_code  # Bad identity
 
 
-# TODO: Remove when workaround is no longer needed
```
it was related to `claassic-proxy` work around but has been removed because now a valid `CN` is expected per https://issues.redhat.com/browse/RHCLOUD-12798

ryandillinfelton replied:
If we expect a valid CN and the workaround is no longer needed then the test for the workaround should be removed.

ryandillinfelton replied:
`test_insights_classic_workaround` and `test_validate_valid_user_identity_no_auth_type` should be deleted

thearifismail commented on tests/helpers/test_utils.py:
```diff
@@ -15,26 +17,28 @@
+USER_IDENTITY = {
+    "account_number": "usraccount",
+    "type": "User",
+    "auth_type": "basic-auth",
```
`enum` values to be implemented in a separate PR.

thearifismail commented on app/queue/queue.py:
```diff
@@ -32,12 +35,7 @@
-    "type": "User",
-    "user": {"email": "tuser@redhat.com", "first_name": "test"},
-}
+SYSTEM_IDENTITY = {"auth_type": "cert-auth", "system": {"cert_type": "system"}, "type": "System"}
```
`enum` values to be implemented in a separate PR.

kruai replied:
Ah - I misunderstood and thought you had created [2 Enums here](https://github.com/RedHatInsights/insights-host-inventory/blob/a5e2f8df186188c18323f0bf691937dc09268130/app/auth/identity.py#L15-L16). Resolving this conversation.

thearifismail commented on tests/helpers/test_utils.py:
```diff
@@ -15,26 +17,28 @@
 SYSTEM_IDENTITY = {
-    "account_number": "test",
+    "account_number": "sysaccount",
     "auth_type": "cert-auth",
```
`enum` values to be implemented in a separate PR.

kruai replied:
(duplicate): I misunderstood and thought you had created [2 Enums here](https://github.com/RedHatInsights/insights-host-inventory/blob/a5e2f8df186188c18323f0bf691937dc09268130/app/auth/identity.py#L15-L16). Resolving this conversation.

thearifismail commented on utils/payloads.py:
```diff
@@ -1,10 +1,22 @@
+    "identity": {
+        "account_number": "sysaccount",
+        "type": "System",
+        "auth_type": "cert-auth",
```
`enum` values to be implemented in a separate PR.

thearifismail commented on tests/helpers/test_utils.py:
```diff
@@ -15,26 +17,28 @@
     "auth_type": "cert-auth",
     "internal": {"auth_time": 6300, "org_id": "3340851"},
-    "system": {"cert_type": "system", "cn": "plxi13y1-99ut-3rdf-bc10-84opf904lfad"},
+    "system": {"cert_type": "system", "cn": "1b36b20f-7fa0-4454-a6d2-008294e06378"},
```
`enum` values to be implemented in a separate PR.

kruai replied:
(duplicate): I misunderstood and thought you had created [2 Enums here](https://github.com/RedHatInsights/insights-host-inventory/blob/a5e2f8df186188c18323f0bf691937dc09268130/app/auth/identity.py#L15-L16). Resolving this conversation.

thearifismail commented on utils/payloads.py:
```diff
@@ -511,11 +525,11 @@ def random_uuid():
 
 
 def build_host_chunk():
-    account = os.environ.get("INVENTORY_HOST_ACCOUNT", "test")
+    account = os.environ.get("INVENTORY_HOST_ACCOUNT", IDENTITY["identity"]["account_number"])
     fqdn = random_uuid()[:6] + ".foo.redhat.com"
     payload = {
         "account": account,
-        "insights_id": random_uuid(),
+        # "insights_id": random_uuid(),
```
Gone!!!

kruai reviewed (changes requested):
Phew, there are hundreds more lines of changes than the last time. I don't think these sweeping changes are the best way to approach this... I've left comments inline with more details

thearifismail commented on tests/helpers/test_utils.py:
```diff
@@ -69,6 +73,18 @@ def set_environment(new_env=None):
 
+# owner_id needed for hosts owned by system accounts.
+def _set_owner_id(values):
+    if "system_profile" not in values.keys():
+        system_profile = {}
+        system_profile["owner_id"] = SYSTEM_IDENTITY["system"]["cn"]
+        values["system_profile"] = system_profile
+    else:
+        if not values["system_profile"].get("owner_id"):
+            values["system_profile"]["owner_id"] = SYSTEM_IDENTITY["system"]["cn"]
```
Done!.  code not pushed yet

thearifismail commented on tests/helpers/test_utils.py:
```diff
@@ -82,6 +98,20 @@ def minimal_host(**values):
+
+    return HostWrapper(data)
+
+
```
`minimal_host_owned_by_system` gone!

thearifismail commented on tests/helpers/db_utils.py:
```diff
@@ -34,9 +39,9 @@ def _clean_tables():
 
 
-def minimal_db_host(**values):
+def minimal_db_host(identity, **values):
```
Done.  Now an `account` is set using `**values`.

kruai commented on tests/test_xjoin.py:
```diff
@@ -1113,28 +1117,32 @@ def test_tags_RBAC_denied(
 def test_bulk_source_header_set_to_db(query_source_xjoin_beta_db, graphql_query_with_response, api_get):
-    response_status, response_data = api_get(HOST_URL, extra_headers={"x-rh-cloud-bulk-query-source": "db"})
+    response_status, response_data = api_get(
+        HOST_URL, SYSTEM_IDENTITY, extra_headers={"x-rh-cloud-bulk-query-source": "db"}
```
Same Q here about using SYSTEM_IDENTITY for these API requests. Again, won't copy/paste it for every instance in this file

thearifismail replied:
there are 4 tests in `tests/test-xjoin.py` which require `SYSTEM_IDENTITY` and 8 more tests using SYSTEM_IDENTITY but can use the default `USER_IDENTITY`.  For better coverage by SYSTEM_IDENTITY, used it in those tests.  There are a lot more tests testing xjoin use the default USER_IDENTITY

kruai replied:
This one is still using `SYSTEM_IDENTITY`

kruai reviewed (changes requested):
I'm requesting many changes here, but most are repeated one-line issues. It's getting closer!

thearifismail commented on app/auth/identity.py:
```diff
@@ -53,6 +56,9 @@ def __init__(self, obj=None, token=None):
                 raise ValueError("Authentication type unknown")
 
+            # Check for bogus parameters
+            validate(self)
```
Done.  Moved it to _get_identity function in queue

thearifismail commented on tests/helpers/api_utils.py:
```diff
@@ -102,31 +99,22 @@ def do_request(
-
-    dict_ = {"identity": identity._asdict()}
+def build_account_auth_header(identity):
+    dict_ = {"identity": Identity(identity)._asdict()}
```
removed the unnecessary Identity creation.

kruai replied:
You didn't remove it, though...

thearifismail replied:
I did but rebasing with the master must have put it back in.  This is what I had and pushing it again.
```
def build_account_auth_header(identity):
    dict_ = {"identity": identity}

    json_doc = json.dumps(dict_)
    auth_header = {"x-rh-identity": b64encode(json_doc.encode())}
    return auth_header
```

thearifismail commented on tests/helpers/api_utils.py:
```diff
@@ -250,20 +238,24 @@ def assert_paginated_response_counts(response_data, expected_per_page, expected_
         with subtests.test(page=page, per_page=per_page):
-            response_status, response_data = api_get(url, query_parameters={"page": page, "per_page": per_page})
+            response_status, response_data = api_get(
+                url, SYSTEM_IDENTITY, query_parameters={"page": page, "per_page": per_page}
```
just to tests pagination using a system account

thearifismail replied:
The tests using this utility require the use of SYSTEM_IDENTITY

kruai replied:
Can you please elaborate about why that identity is required? It doesn't make sense to me for our API tests to use `SYSTEM_IDENTITY`

thearifismail replied:
Gone!

thearifismail commented on tests/test_synchronizer.py:
```diff
@@ -18,7 +19,7 @@ def test_synchronize_host_event(
 
     host = minimal_db_host(stale_timestamp=staleness_timestamps["culled"], reporter="some reporter")
-    created_host = db_create_host(host)
+    created_host = db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, it's optional (see previous)

thearifismail replied:
Gone!

thearifismail commented on tests/test_synchronizer.py:
```diff
@@ -43,7 +44,7 @@ def test_synchronize_host_event(
     host_count = 25
 
-    db_create_multiple_hosts(how_many=host_count)
+    db_create_multiple_hosts(USER_IDENTITY, how_many=host_count)
```
identity is required when hosts info is passed to db_create_multiple_hosts()

kruai replied:
nope, it's optional (see previous)

thearifismail replied:
Gone!

thearifismail commented on tests/test_system_profile.py:
```diff
@@ -20,18 +20,23 @@
     system_profile = valid_system_profile()
+    system_profile["owner_id"] = OWNER_ID
     host = minimal_host(system_profile=system_profile)
     created_host = mq_create_or_update_host(host)
 
     url = build_system_profile_url(host_list_or_id=created_host.id)
 
-    response_status, response_data = api_get(url)
+    response_status, response_data = api_get(url, SYSTEM_IDENTITY)
```
The tests using `mq_create_or_update_host` fixture require `api_get(url, SYSTEM_IDENTITY)` because the way the fixture was updated to test system_identity.

kruai replied:
If the fixture being modified requires a bunch of tests to be changed, then something is likely wrong with the change to the fixture. If every `api_get()` test is changed to use SYSTEM_IDENTITY, then we are only testing the scenarios where a system is making GET calls (I don't even think that happens).

thearifismail replied:
SYSTEM_IDENTITY gone from all `api_get() calls` except the test IS about using the SYSTEM_IDENTITY

thearifismail commented on tests/test_deduplication.py:
```diff
@@ -58,7 +59,7 @@ def test_find_host_using_subscription_manager_id_match(db_create_host):
 
     host = minimal_db_host(canonical_facts=canonical_facts)
-    created_host = db_create_host(host)
+    created_host = db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_deduplication.py:
```diff
@@ -43,7 +44,7 @@ def test_find_host_using_insights_id_match(db_create_host):
 
     host = minimal_db_host(canonical_facts=canonical_facts)
-    created_host = db_create_host(host)
+    created_host = db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_deduplication.py:
```diff
@@ -27,7 +28,7 @@ def test_find_host_using_superset_canonical_fact_match(db_create_host):
 
     host = minimal_db_host(canonical_facts=canonical_facts)
-    created_host = db_create_host(host)
+    created_host = db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_deduplication.py:
```diff
@@ -3,14 +3,15 @@
 
     host = minimal_db_host(canonical_facts=canonical_facts)
-    created_host = db_create_host(host)
+    created_host = db_create_host(USER_IDENTITY, host)
```
test_find_host_using_superset_canonical_fact_match

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_culling.py:
```diff
@@ -342,7 +347,7 @@ def test_non_culled_host_is_not_removed(
     ):
         host = minimal_db_host(stale_timestamp=stale_timestamp, reporter="some reporter")
-        created_host = db_create_host(host)
+        created_host = db_create_host(USER_IDENTITY, host)
```
test_find_host_using_superset_canonical_fact_match

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_api_hosts_get.py:
```diff
@@ -33,13 +34,16 @@
 
 
 def test_query_all(mq_create_three_specific_hosts, api_get, subtests):
     created_hosts = mq_create_three_specific_hosts
     expected_host_list = build_expected_host_list(created_hosts)
 
-    response_status, response_data = api_get(HOST_URL)
+    response_status, response_data = api_get(HOST_URL, SYSTEM_IDENTITY)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
Actually the Identity param is optional, but that's not what I'm asking here. I'm asking why all of these API tests have been changed to use `SYSTEM_IDENTITY`, when we almost always get User identities in API calls

thearifismail commented on tests/test_api_hosts_update.py:
```diff
@@ -216,7 +217,7 @@ def test_patch_produces_update_event_no_request_id(
 ):
     host = db_host()
-    created_host = db_create_host(host)
+    created_host = db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
No, the Identity is optional - see my previous comments

thearifismail commented on tests/test_api_hosts_update.py:
```diff
@@ -240,7 +241,7 @@ def test_patch_produces_update_event_with_request_id(
 
     host = db_host()
-    created_host = db_create_host(host)
+    created_host = db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
No, the Identity is optional - see my previous comments

thearifismail commented on tests/test_api_hosts_update.py:
```diff
@@ -260,7 +261,7 @@ def test_patch_produces_update_event_no_insights_id(
     del host.canonical_facts["insights_id"]
 
-    created_host = db_create_host(host)
+    created_host = db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
No, the Identity is optional - see my previous comments

thearifismail commented on tests/test_api_hosts_get.py:
```diff
@@ -33,13 +34,16 @@
 from tests.helpers.test_utils import minimal_host
 from tests.helpers.test_utils import now
+from tests.helpers.test_utils import SYSTEM_IDENTITY
+
+# from tests.helpers.test_utils import USER_IDENTITY
```
Gone.

thearifismail commented on tests/test_api_hosts_delete.py:
```diff
@@ -253,7 +262,7 @@ def test_delete_stops_after_kafka_producer_error(
     event_producer._kafka_producer.send.side_effect = send_side_effects
 
-    hosts = db_create_multiple_hosts(how_many=3)
+    hosts = db_create_multiple_hosts(USER_IDENTITY, how_many=3)
```
identity is required when hosts info is passed to db_create_multiple_hosts()

kruai replied:
No, it's optional. Same answer as above

thearifismail commented on tests/test_api_hosts_delete.py:
```diff
@@ -234,7 +243,7 @@ def test_delete_hosts_chunk_size(
     mocker.patch("api.host._get_host_list_by_id_list", query_wraper.mock_get_host_list_by_id_list)
 
-    hosts = db_create_multiple_hosts(how_many=2)
+    hosts = db_create_multiple_hosts(USER_IDENTITY, how_many=2)
```
identity is required when hosts info is passed to db_create_multiple_hosts()

kruai replied:
No, it's optional. Same answer as above

thearifismail commented on tests/test_api_hosts_delete.py:
```diff
@@ -151,7 +158,7 @@ def test_delete_when_all_hosts_are_deleted(event_producer_mock, db_create_multip
 
 def test_delete_when_some_hosts_is_deleted(event_producer_mock, db_create_multiple_hosts, api_delete_host, mocker):
-    hosts = db_create_multiple_hosts(how_many=2)
+    hosts = db_create_multiple_hosts(USER_IDENTITY, how_many=2)
```
identity is required when hosts info is passed to db_create_multiple_hosts()

kruai replied:
No, it's optional. Same answer as above

thearifismail replied:
Gone!

thearifismail commented on tests/test_api_hosts_delete.py:
```diff
@@ -122,21 +127,23 @@ def test_create_then_delete_check_metadata(event_datetime_mock, event_producer_m
 
 def test_delete_when_all_hosts_are_deleted(event_producer_mock, db_create_multiple_hosts, api_delete_host, mocker):
-    hosts = db_create_multiple_hosts(how_many=2)
+    hosts = db_create_multiple_hosts(USER_IDENTITY, how_many=2)
```
identity is required when hosts info is passed to db_create_multiple_hosts()

kruai replied:
No, it's optional. Same answer as above

thearifismail replied:
Gone!

thearifismail commented on tests/test_api_hosts_delete.py:
```diff
@@ -93,7 +96,7 @@ def test_create_then_delete_without_insights_id(
     del host.canonical_facts["insights_id"]
 
-    db_create_host(host)
+    db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
No, the Identity is an optional parameter. To pass the `host` param and not the `identity` one, you can just use `db_create_host(host=host)`

thearifismail replied:
Gone!

thearifismail commented on tests/fixtures/api_fixtures.py:
```diff
@@ -13,111 +13,44 @@ def flask_client(flask_app):
-            identity_type=identity_type,
-        )
+    def _api_post(url, host_data, identity=USER_IDENTITY, query_parameters=None, extra_headers=None):
+        return do_request(flask_client.post, url, identity, host_data, query_parameters, extra_headers)
```
agree

thearifismail commented on tests/helpers/api_utils.py:
```diff
@@ -250,20 +238,24 @@ def assert_paginated_response_counts(response_data, expected_per_page, expected_
     for parameter, invalid_value in product(("per_page", "page"), ("-1", "0", "notanumber")):
         with subtests.test(parameter=parameter, invalid_value=invalid_value):
-            response_status, response_data = api_get(url, query_parameters={parameter: invalid_value})
+            response_status, response_data = api_get(url, SYSTEM_IDENTITY, query_parameters={parameter: invalid_value})
```
Removed though using both identity types worked.

thearifismail commented on tests/helpers/api_utils.py:
```diff
@@ -250,20 +238,24 @@ def assert_paginated_response_counts(response_data, expected_per_page, expected_
     non_existent_page = expected_total + 1
-    response_status, response_data = api_get(url, query_parameters={"page": non_existent_page, "per_page": 1})
+    response_status, response_data = api_get(
+        url, SYSTEM_IDENTITY, query_parameters={"page": non_existent_page, "per_page": 1}
```
Removed though using both identity types worked.

thearifismail commented on tests/helpers/test_utils.py:
```diff
@@ -69,6 +73,12 @@ def set_environment(new_env=None):
     patched_dict.stop()
 
 
+# owner_id needed for hosts owned by system accounts.
+def _set_owner_id(values):
+    values.setdefault("system_profile", {}).setdefault("owner_id", SYSTEM_IDENTITY["system"]["cn"])
+    return values
+
+
```
Gone!

thearifismail commented on tests/test_models.py:
```diff
@@ -358,7 +378,7 @@ def test_update_per_reporter_staleness(db_create_host, models_datetime_mock):
         {"fqdn": "fqdn"}, display_name="display_name", reporter="puptoo", stale_timestamp=puptoo_stale_timestamp
     )
-    existing_host = db_create_host(input_host)
+    existing_host = db_create_host(USER_IDENTITY, input_host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_models.py:
```diff
@@ -342,7 +362,7 @@ def test_create_host_sets_per_reporter_staleness(db_create_host, models_datetime
         {"fqdn": "fqdn"}, display_name="display_name", reporter="puptoo", stale_timestamp=stale_timestamp
     )
-    created_host = db_create_host(input_host)
+    created_host = db_create_host(USER_IDENTITY, input_host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_models.py:
```diff
@@ -333,7 +353,7 @@ def test_host_model_constraints(field, value, db_create_host):
 
     with pytest.raises(DataError):
-        db_create_host(host)
+        db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_models.py:
```diff
@@ -308,7 +328,7 @@ def test_host_model_timestamp_timezones(db_create_host):
     )
 
-    db_create_host(host)
+    db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_models.py:
```diff
@@ -287,7 +307,7 @@ def test_host_model_updated_timestamp(db_create_host):
 
     before_insert_commit = now()
-    db_create_host(host)
+    db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_models.py:
```diff
@@ -269,7 +289,7 @@ def test_host_model_default_timestamps(db_create_host):
 
     before_commit = now()
-    db_create_host(host)
+    db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_models.py:
```diff
@@ -255,7 +275,7 @@ def test_host_model_default_id(db_create_host):
         stale_timestamp=now(),
     )
-    db_create_host(host)
+    db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_models.py:
```diff
@@ -241,7 +261,7 @@ def test_host_model_assigned_values(db_create_host, db_get_host):
 
     inserted_host = Host(**values)
-    db_create_host(inserted_host)
+    db_create_host(USER_IDENTITY, inserted_host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on tests/test_deduplication.py:
```diff
@@ -70,7 +71,7 @@ def test_find_host_using_elevated_ids_match(db_create_host, host_create_order, e
     for host_canonical_facts in host_create_order:
         host = minimal_db_host(canonical_facts=hosts_canonical_facts[host_canonical_facts])
-        created_host = db_create_host(host)
+        created_host = db_create_host(USER_IDENTITY, host)
```
identity is required when a host info is passed to db_create_host()

kruai replied:
nope, optional (see previous)

thearifismail commented on app/queue/queue.py:
```diff
@@ -136,11 +177,15 @@ def handle_message(message, event_producer):
+    identity = _get_identity(host, platform_metadata)
+
+    if host.get("account") != identity.account_number:
+        raise "The account number in identity does not match the number in the host."
```
Now raises `ValidationException`

thearifismail commented on tests/test_rbac.py:
```diff
@@ -2,6 +2,8 @@
 
 from tests.helpers.api_utils import build_hosts_url
 
+# from tests.helpers.test_utils import USER_IDENTITY
+
```
Gone.

thearifismail commented on tests/test_tags.py:
```diff
@@ -26,7 +27,7 @@ def test_get_tags_of_multiple_hosts(mq_create_four_specific_hosts, api_get, subt
     expected_response = {host.id: host.tags for host in created_hosts}
 
     url = build_host_tags_url(host_list_or_id=created_hosts, query="?order_by=updated&order_how=ASC")
-    response_status, response_data = api_get(url)
+    response_status, response_data = api_get(url, SYSTEM_IDENTITY)
```
The tests using mq_create_four_specific_hosts fixture require api_get(url, SYSTEM_IDENTITY) because the way the fixture was updated to test system_identity.

thearifismail replied:
SYSTEM_IDENTITY gone from all `api_get() calls` except the test IS about using the SYSTEM_IDENTITY

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -1058,10 +1053,124 @@ def test_handle_message_side_effect(mocker, flask_app):
+
+    message = wrap_message(host.data(), "add_host", get_platform_metadata_with_system_identity())
+
+    with pytest.raises(InventoryException):
```
Changed to pytest.raises(ValidationException)

thearifismail commented on tests/test_culling.py:
```diff
@@ -34,18 +36,18 @@ def test_with_stale_timestamp(mq_create_or_update_host, api_get):
     updated_host = mq_create_or_update_host(host)
     assert_system_culling_data(updated_host.data(), stale_timestamp, reporter)
 
-    response_status, response_data = api_get(HOST_URL)
+    response_status, response_data = api_get(HOST_URL, SYSTEM_IDENTITY)
```
SYSTEM_IDENTITY is required when a host is created using `mq_create_or_update_host`

kruai replied:
The `SYSTEM_IDENTITY` requirement is an issue you created with this line: https://github.com/RedHatInsights/insights-host-inventory/blob/4d5e98f217381a3d728add3bd85bb08569d117f2/tests/fixtures/mq_fixtures.py#L31
API calls are almost always users, so it doesn't make sense to change all these tests to test system identities. IMO,  `mq_create_or_update_host` needs to be updated with another optional param (either Identity or account number).

thearifismail commented on tests/test_api_hosts_delete.py:
```diff
@@ -205,19 +212,21 @@ def test_delete_host_with_RBAC_denied(
 
             host = db_create_host()
 
-            response_status, response_data = api_delete_host(host.id, identity_type="User")
+            response_status, response_data = api_delete_host(host.id, SYSTEM_IDENTITY)
 
-            assert_response_status(response_status, 403)
+            assert_response_status(response_status, 404)
```
The test attempts to delete a host not created by it, so the host is not found that's why the error 404 is retuned by
`        if not query.count():
            flask.abort(status.HTTP_404_NOT_FOUND)
`

kruai replied:
That's not what this is testing, though. This is an RBAC test, which is attempting to validate that the RBAC checks restrict permissions correctly; this test isn't trying to test anything related to `owner_id`

thearifismail replied:
Fixed.

thearifismail commented on tests/test_mq_service.py:
```diff
@@ -292,6 +271,23 @@ def test_add_host_with_system_profile(event_datetime_mock, mq_create_or_update_h
+    )
+
+    with pytest.raises(InventoryException):
+        key, event, headers = mq_create_or_update_host(host, return_all_data=True)
```
for now changed it to pytest.raise(ValidationException)

thearifismail replied:
Changed to pytest.raises(ValidationException) and asserted for the exact message "The owner in host does not match the owner in identity"

kruai replied:
perfect!

kruai reviewed (changes requested):
I responded to a lot of your comments, though I eventually stopped bc I ended up just copy/pasting

ryandillinfelton commented on tests/test_xjoin.py:
```diff
@@ -1694,7 +1689,7 @@ def test_query_system_profile_sap_system_insights_classic_system_identity(
 ):
     url = build_system_profile_sap_system_url()
 
-    response_status, response_data = api_get(url, identity_type="Insights_Classic_System")
```
These test that use the insight_classic_system identity pass but no longer tests what they're supposed to. If we no longer need special handling for insights_classic_system identities we should just remove these tests, if we do still need special handling these tests need to be restored to using insights_classic_system identities.

thearifismail replied:
The `Insights_Classic_Identity` is being fix by https://issues.redhat.com/browse/RHCLOUD-12798.  For now leaving these test to not forget Insights-Classic until the fix is available.  Very likely will be removed by separate PR.

kruai replied:
@thearifismail Isn't that Jira unrelated to Ryan's comment? These "Insights_Classic_System" tests are no longer testing what they were created for

ryandillinfelton commented on tests/test_tags.py:
```diff
@@ -325,17 +326,18 @@ def test_get_host_tags_with_RBAC_denied(subtests, mocker, db_create_host, api_ge
     get_rbac_permissions_mock = mocker.patch("lib.middleware.get_rbac_permissions")
 
-    created_hosts = mq_create_four_specific_hosts
+    created_hosts = [db_create_host()]
```
Why change this from multiple hosts?

thearifismail replied:
Back to using multiple hosts.

kruai commented on tests/test_culling.py:
```diff
@@ -101,7 +102,7 @@ def test_get_multiple_states(mq_create_hosts_in_all_states, api_get):
 def test_get_hosts_list_default_ignores_culled(mq_create_hosts_in_all_states, api_get):
     created_hosts = mq_create_hosts_in_all_states
 
-    response_status, response_data = api_get(HOST_URL)
+    response_status, response_data = api_get(HOST_URL, SYSTEM_IDENTITY)
```
All the tests in this file are still using `SYSTEM_IDENTITY` for API request testing

thearifismail replied:
Fixed.

kruai commented on tests/test_system_profile.py:
```diff
@@ -185,7 +190,7 @@ def test_get_system_profile_of_host_that_does_not_exist(api_get):
     expected_total = 0
     host_id = generate_uuid()
 
-    response_status, response_data = api_get(f"{HOST_URL}/{host_id}/system_profile")
+    response_status, response_data = api_get(f"{HOST_URL}/{host_id}/system_profile", SYSTEM_IDENTITY)
```
This one is still using `SYSTEM_IDENTITY`

thearifismail replied:
Fixed.

kruai commented on tests/test_system_profile.py:
```diff
@@ -195,7 +200,7 @@ def test_get_system_profile_of_host_that_does_not_exist(api_get):
 
 @pytest.mark.parametrize("invalid_host_id", ["notauuid", "922680d3-4aa2-4f0e-9f39-38ab8ea318bb,notuuid"])
 def test_get_system_profile_with_invalid_host_id(api_get, invalid_host_id):
-    response_status, response_data = api_get(f"{HOST_URL}/{invalid_host_id}/system_profile")
+    response_status, response_data = api_get(f"{HOST_URL}/{invalid_host_id}/system_profile", SYSTEM_IDENTITY)
```
This one is still using `SYSTEM_IDENTITY`

thearifismail replied:
Fixed.

kruai commented on tests/test_xjoin.py:
```diff
@@ -1113,28 +1117,32 @@ def test_tags_RBAC_denied(
 def test_bulk_source_header_set_to_xjoin(query_source_xjoin_beta_db, graphql_query_with_response, api_get):
-    response_status, response_data = api_get(HOST_URL, extra_headers={"x-rh-cloud-bulk-query-source": "xjoin"})
+    response_status, response_data = api_get(
+        HOST_URL, SYSTEM_IDENTITY, extra_headers={"x-rh-cloud-bulk-query-source": "xjoin"}
```
It looks like a lot of the tests in this file are also still using `SYSTEM_IDENTITY` for API calls

thearifismail replied:
Fixed.

kruai reviewed (changes requested):
A whole lot of the API calls in the tests are still using `SYSTEM_IDENTITY`...

kruai reviewed (commented):
`_db_create_host()`'s identity param should probably default to the mock System Identity bc systems are the only ones creating hosts at this point, but you've got tests that use both identities, so it's not critical IMO. I second Ryan's requested changes, but I think we're almost there

ryandillinfelton reviewed (approved):
LGTM :rocket:

kruai reviewed (approved):
:+1:

tahmidefaz reviewed (approved):
yolo!

thearifismail commented:
this PR better be YOLO
